sasdata/data_util/manipulations.py

```diff
@@ -303,13 +302,16 @@ def get_bin_index(self, value):
         bin = floor(N * (log(x) - log(min)) / (log(max) - log(min)))
         """
         if self.base:
-            temp_x = self.n_bins * (math.log(value, self.base) - math.log(self.min, self.base))
+            temp_x = math.log(value, self.base) - math.log(self.min, self.base)
             temp_y = math.log(self.max, self.base) - math.log(self.min, self.base)
         else:
-            temp_x = self.n_bins * (value - self.min)
+            temp_x = value - self.min
             temp_y = self.max - self.min
+        # Fixing an issue where certain angular coordinate values give bad bin values
+        if np.fabs(temp_x) > np.fabs(temp_y):
+            temp_x = normalize_angle(temp_x)
         # Bin index calulation
```

Comment:
I don't think this is the appropriate fix is it?  this is a generalized binning class that returns a bin value given the min and max of a range, the number of bins being used, and a value whose bin is to be computed as I understand the class? In which case it should not know anything about the type of object that is being binned.  Thus `normalize_angle`, which assumes the range (min, max ) being binned is an angular range should not be in this class (for example boxslicer does not have angular ranges). If we need to trap such an error nicely maybe just return an error log message?  I would expect that the calling class making use of this binning class should be responsible to first call `normalize_angle` prior to sending to binning if that class is sending an angular range?

Reply:
The class may not be aware of the type of object being binned, but based on how it's called I'm fairly certain the only time |`temp_x`| can ever be greater than |`temp_y`| would be when it's working with angular values. So it shouldn't cause problems for non-angular slicers unless the `value` parameter it's called with makes no sense (`value` would have to be greater than `self.max`). I see your point about ensuring the calling class normalises the angles instead, but for the bug this PR set out to solve, that isn't enough.

Let me explain the original issue a bit more thoroughly. For whatever confusing reason, `manipulations.py` is set up to have angles measured anticlockwise from the negative x-axis. So 0 - pi/2 is in the third quadrant, pi/2 - pi is in the fourth, pi - 3pi/2 is in the first quadrant and 3pi/2 - 2pi in the second. When `get_bin_index` is called for angular averaging, it is done in a loop with the `value` argument gradually moving in the anticlockwise direction, from `phi_min` to `phi_max`. Issue #40 occurs when `phi_min` is in the second quadrant, and `phi_max` is in the third. When `value` is in the second quadrant, `(value - min)/(max - min)` is between -1 and 0, weird but not problematic. When `value` is in the third quadrant however, `(value - min)/(max - min)` > 1. This is what results in the out of range bin indices. I tried experimenting with setting `min` and `max` in `Binning()` so that `max` is always greater, but it didn't help.

Reply:
I'm planning to do a full refactor of the binning methods used in `manipulations.py` once I've finished with these bugs, with the chief goal of introducing working fractional binning.  Ultimately I think circular/angular binning and square binning are different enough that they should be handled by different methods, but that can be saved for the re-work. For now, I'd like to keep this `normalize_angle()` call in place, especially since it isn't actually causing any issues for other types of averaging.

Reply:
I am confused by the suggestion that the binning class needs refactoring. It is actually a very general class that takes a range from x to y, a number of bins desired and a value you want to know the bin number of an returns it either in linear or log? I don't see how that is causing any problem? There are certainly questions I can imagine with dealing with how one weights a pixel that crosses a bin boundary but that is all done in the `get_pixel_fraction` function I thought? If not then it should be.

On the other hand, if the binning code DOES get thrown away, it seems to me that is an even stronger argument for putting the code where I believe it belongs: the sector class just before calling binning? That way your solution for dealing with sectors that straddle 0, whatever it is, is preserved and does not have to be redesigned into the new binning class?

Which brings me to my issue with the current fix, as suggested in my comment to issue #40, the error is not really coming from the `Binning` class. that class is doing what it should. the problem is that you are calling it with incorrect values (you are giving it a min > max basically).  That should be addressed it seems to me in the code that knows about that fact. I believe the place to address the problem is in the `_avg` method of the `_Sector` class.

In fact there is already a TON of code in that method dealing with **exactly** this straddling zero problem for other sector type slices.. and I mean a lot... as it should be I would argue (though see caveat at end)

Basically the fix should go around [line 889 ](https://github.com/SasView/sasdata/blob/40-return-correct-bin-index-for-angular-coordinates/sasdata/data_util/manipulations.py#L887-L894) where the binning object is created

```python
        #  set up the bins by creating a binning object
        if run.lower() == 'phi':
            binning = Binning(self.phi_min, self.phi_max, self.nbins, self.base)
        elif self.fold:
            binning = Binning(self.r_min, self.r_max, self.nbins, self.base)
        else:
            binning = Binning(-self.r_max, self.r_max, self.nbins, self.base)
```

and then in the averaging loop when retrieving the binning index aroun [line 950](https://github.com/SasView/sasdata/blob/40-return-correct-bin-index-for-angular-coordinates/sasdata/data_util/manipulations.py#L950-L955) 

```python
            # Get the binning index
            if run.lower() == 'phi':
                i_bin = binning.get_bin_index(phi_value)
            else:
                i_bin = binning.get_bin_index(q_value)

            # Take care of the edge case at phi = 2pi.
            if i_bin == self.nbins:
                i_bin = self.nbins - 1
```
Unfortunately looking at the current code I have realized this evening that I'm pretty sure that math is wrong. everything will be put in the lower index bins unless I'm doing the math wrong? I think the easiest way to deal with this correctly if one wants to stick to the 0<phi<2pi  range that all this code seems to like, would be to implement a rotation of the wedge such that phi_min is 0.  Speaking with @lucas-wilkins just now, assuming we don't care about 0<phi<2pi then rotating the wedge by 2pi would be even easier to implement code wise (just add 2pi to each of phi_min, phi_max, and phi_value).  given the transformation is just to send to binning I would agree that there is no need to try to keep 0<phi<2pi. 

In fact it is even simpler I realize, which @lucas-wilkins may have already known. All you need to do is add 2pi to phi_max if phi_min > phi_max and add 2pi to phi_value if phi_min > phi_value 

so my thinking is that the end result would look something like this?

on line 889 add
```phython
if phi_min > phi_max: phi_max =+ 2*np.py
```
to yield
```python
        #  set up the bins by creating a binning object
        if run.lower() == 'phi':
            if phi_min > phi_max: phi_max =+ 2*np.py
            binning = Binning(self.phi_min, self.phi_max, self.nbins, self.base)
        elif self.fold:
            binning = Binning(self.r_min, self.r_max, self.nbins, self.base)
        else:
            binning = Binning(-self.r_max, self.r_max, self.nbins, self.base)
```
and on line 953 (was 952) add
```python
if phi_min>phi_value: phi_value =+2*np.py
```
```python
            # Get the binning index
            if run.lower() == 'phi':
                if phi_min>phi_value: phi_value =+2*np.py
                i_bin = binning.get_bin_index(phi_value)
            else:
                i_bin = binning.get_bin_index(q_value)

            # Take care of the edge case at phi = 2pi.
            if i_bin == self.nbins:
                i_bin = self.nbins - 1
```
Having now gone through this whole exercise, I'm wondering if this method would not allow one to clean up a bunch of the lines of code currently trying to deal with this problem?

Reply:
Just to be clear - that last bit is way out of scope for this PR. Just an observation that may be worth another look at some point?

Reply:
When I talk about refactoring the binning, I don't just mean refactoring the class `Binning`, I mean a redesign of how bins work everywhere in `manipulations.py` to get fractional binning working. Somebody has started working on this with `get_pixel_fraction`, but it's not implemented yet. So, I'll be spending a lot more time thinking about these problems in the near future.

I think you're right that it shouldn't be `Binning.get_bin_index`'s job to sort out this zero-straddling error. I wanted to get a quick fix in so that wedge slicer could be made fully functional, and the approach I've gone with so far was just the first thing I found that worked. I'll try out the fix you've suggested, but the commit history in this branch is getting pretty messy. Perhaps I should just make a new branch addressing Issue #40 and scrap this one.

Reply:
I've created a new branch and new PR which implements an alternative fix which I think everyone involved will be much happier with. You can find it here: #42